src/GafferSceneUI/SceneGadget.cpp

```diff
@@ -290,6 +429,55 @@ bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, Gaff
 }
 
 bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, GafferScene::ScenePlug::ScenePath &path, V3f &hitPoint ) const
+{
+	if( m_updateErrored )
+	{
+		return false;
+	}
+
+	float depth = std::numeric_limits<float>::max();
+	bool hit = openGLObjectAt( lineInGadgetSpace, path, depth );
+
+	auto viewportGadget = ancestor<ViewportGadget>();
+	if( m_outputBuffer )
+	{
+		const V2f rasterPosition = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p1, this );
+		float bufferDepth;
+		if( uint32_t id = m_outputBuffer->idAt( rasterPosition / viewportGadget->getViewport(), bufferDepth ) )
+		{
+			if( bufferDepth < depth )
+			{
+				if( auto bufferPath = m_controller->pathForID( id ) )
+				{
+					depth = bufferDepth;
+					path = *bufferPath;
+					hit = true;
+				}
+			}
+		}
+	}
+
+	if( !hit )
+	{
+		return false;
+	}
+
+	V3f viewDir;
+	const M44f cameraWorldTransform = viewportGadget->getCameraTransform();
+	const M44f cameraTransform = cameraWorldTransform * fullTransform().inverse();
+	cameraTransform.multDirMatrix( V3f( 0.0f, 0.0f, -1.0f ), viewDir );
+
+	const V3f traceDir = lineInGadgetSpace.normalizedDirection();
+
+	depth /= max( 0.00001f, viewDir.dot( traceDir ) );
+
+	const V3f origin = V3f( 0.0f ) * cameraTransform;
+	hitPoint = origin + ( traceDir * depth );
+
+	return true;
+}
+
+bool SceneGadget::openGLObjectAt( const IECore::LineSegment3f &lineInGadgetSpace, GafferScene::ScenePlug::ScenePath &path, float &depth ) const
```

Comment:
Logic seems sound, but I find the names a little misleading:  openGLObjectAt and objectAt has such similar names that it feels weird they don't have the same return values.  Because openGLObjectAt is private, it doesn't really matter.